File: sdk-python/obsvr/strict_receipt_verify.py
"""Offline verification for strict receipts and receipt chains."""

from __future__ import annotations

import base64
import binascii
import hashlib
from typing import Any, Dict, List, Literal, Optional

from .strict_receipt import (
    STRICT_RECEIPT_BODY_DOMAIN,
    STRICT_RECEIPT_ENVELOPE_SCHEMA,
    StrictReceiptValidationError,
    _canonical_json_for_hash,
    _u64,
    build_strict_receipt_body,
    strict_receipt_key_id,
    strict_receipt_signature_preimage,
)

_HEX = frozenset("0123456789abcdef")
StrictKeyTrust = Literal[
    "pinned", "registered", "revoked", "self_asserted", "unknown"
]


def _record(value: Any) -> Optional[Dict[str, Any]]:
    return value if isinstance(value, dict) else None


def _exact(value: Dict[str, Any], allowed: set[str]) -> bool:
    return set(value).issubset(allowed)


def _lower_hex(value: Any, length: int) -> bool:
    return (
        isinstance(value, str)
        and len(value) == length
        and all(char in _HEX for char in value)
    )


def _key_id(value: Any) -> bool:
    return isinstance(value, str) and value.startswith("sha256:") and _lower_hex(
        value[7:], 64
    )


def _decode_public_key(value: Any) -> Optional[bytes]:
    if not isinstance(value, str):
        return None
    try:
        raw = base64.b64decode(value, validate=True)
        return raw if len(raw) == 32 and base64.b64encode(raw).decode() == value else None
    except (binascii.Error, ValueError):
        return None


def verify_strict_receipt(
    value: Any, *, pinned_public_key_b64: Optional[str] = None
) -> Dict[str, Any]:
    envelope = _record(value)
    body = _record(envelope.get("body")) if envelope else None
    signature = _record(envelope.get("signature")) if envelope else None
    schema_valid = bool(
        envelope is not None
        and _exact(
            envelope,
            {"schema", "body", "receipt_hash", "signature", "public_key_b64"},
        )
        and envelope.get("schema") == STRICT_RECEIPT_ENVELOPE_SCHEMA
        and body is not None
        and _lower_hex(envelope.get("receipt_hash"), 64)
        and signature is not None
        and _exact(signature, {"algorithm", "key_id", "value"})
        and signature.get("algorithm") == "Ed25519"
        and _key_id(signature.get("key_id"))
        and _lower_hex(signature.get("value"), 128)
        and (
            "public_key_b64" not in envelope
            or isinstance(envelope["public_key_b64"], str)
        )
    )

    semantic_valid = False
    if body is not None:
        try:
            build_strict_receipt_body(body)
            semantic_valid = True
        except StrictReceiptValidationError:
            pass

    hash_valid = False
    if body is not None and envelope is not None and isinstance(
        envelope.get("receipt_hash"), str
    ):
        try:
            canonical = _canonical_json_for_hash(body).encode("utf-8")
            actual = hashlib.sha256(
                STRICT_RECEIPT_BODY_DOMAIN
                + b"\x00"
                + _u64(len(canonical))
                + canonical
            ).hexdigest()
            hash_valid = actual == envelope["receipt_hash"]
        except Exception:
            pass

    pin_supplied = pinned_public_key_b64 is not None
    pinned = _decode_public_key(pinned_public_key_b64)
    embedded = _decode_public_key(envelope.get("public_key_b64") if envelope else None)
    raw_key = pinned if pin_supplied else embedded
    key_trust: StrictKeyTrust = (
        "pinned" if pin_supplied else "self_asserted" if embedded else "unknown"
    )
    initiator = _record(body.get("initiator")) if body else None
    body_key_id = initiator.get("key_id") if initiator else None
    signature_key_id = signature.get("key_id") if signature else None
    fingerprint = strict_receipt_key_id(raw_key) if raw_key is not None else None
    identity_valid = bool(
        isinstance(body_key_id, str)
        and isinstance(signature_key_id, str)
        and fingerprint is not None
        and body_key_id == signature_key_id == fingerprint
    )

    signature_valid = False
    if (
        raw_key is not None
        and isinstance(signature_key_id, str)
        and envelope is not None
        and isinstance(envelope.get("receipt_hash"), str)
        and signature is not None
        and isinstance(signature.get("value"), str)
    ):
        try:
            from .policy_verify import _resolve_backend

            backend = _resolve_backend()
            if backend is not None:
                signature_valid = bool(
                    backend(
                        raw_key,
                        strict_receipt_signature_preimage(
                            signature_key_id, envelope["receipt_hash"]
                        ),
                        bytes.fromhex(signature["value"]),
                    )
                )
        except Exception:
            pass

    return {
        "schema_valid": schema_valid,
        "hash_valid": hash_valid,
        "signature_valid": signature_valid,
        "semantic_valid": semantic_valid,
        "identity_binding_valid": identity_valid,
        "key_trust": key_trust,
    }


def _same_action(left: Dict[str, Any], right: Dict[str, Any]) -> bool:
    return all(
        left.get(field) == right.get(field)
        for field in ("action_id", "kind", "name", "arguments_hash", "target")
    )


def _compatible_resolution(suspension: Dict[str, Any], method: str) -> bool:
    if suspension.get("type") == "approval":
        return method in (
            "approval_granted",
            "approval_denied",
            "expired",
            "cancelled",
        )
    return method in ("context_supplied", "expired", "cancelled")


def verify_strict_receipt_chain(
    envelopes: List[Any], *, pinned_public_key_b64: Optional[str] = None
) -> Dict[str, Any]:
    if not envelopes:
        return {"valid": False, "errors": ["empty_chain"]}
    errors: List[str] = []
    by_hash: Dict[str, Dict[str, Any]] = {}
    resolved: set[str] = set()
    session = None
    previous = None
    for index, candidate in enumerate(envelopes):
        current = _record(candidate)
        body = _record(current.get("body")) if current else None
        receipt_id = body.get("receipt_id", f"index-{index}") if body else f"index-{index}"
        axes = verify_strict_receipt(
            candidate, pinned_public_key_b64=pinned_public_key_b64
        )
        for axis, label in (
            ("schema_valid", "receipt_schema_invalid"),
            ("semantic_valid", "receipt_semantic_invalid"),
            ("hash_valid", "receipt_hash_invalid"),
            ("signature_valid", "receipt_signature_invalid"),
            ("identity_binding_valid", "receipt_identity_invalid"),
        ):
            if not axes[axis]:
                errors.append(f"{label}:{receipt_id}")
        if current is None or body is None or not all(
            axes[key]
            for key in (
                "schema_valid",
                "semantic_valid",
                "hash_valid",
                "signature_valid",
                "identity_binding_valid",
            )
        ):
            continue
        body = build_strict_receipt_body(body)
        if session is None:
            session = body["session_id"]

        if body["sequence"] != index + 1:
            errors.append(f"sequence_order_invalid:{receipt_id}")
        if body["session_id"] != session:
            errors.append(f"session_mismatch:{receipt_id}")
        if previous is not None:
            previous_body = previous["body"]
            if body["previous_receipt_hash"] != previous["receipt_hash"]:
                errors.append(f"previous_hash_mismatch:{receipt_id}")
            if body["timestamp_ms"] < previous_body["timestamp_ms"]:
                errors.append(f"timestamp_regression:{receipt_id}")

        resolution = body.get("resolution")
        if body["record_type"] == "resolution" and resolution:
            if resolution["resolves_receipt_hash"] in resolved:
                errors.append(f"duplicate_resolution:{receipt_id}")
            else:
                resolved.add(resolution["resolves_receipt_hash"])
            prior = by_hash.get(resolution["resolves_receipt_hash"])
            if prior is None:
                errors.append(f"resolution_reference_invalid:{receipt_id}")
            else:
                prior_body = prior["body"]
                suspension = prior_body.get("suspension")
                if prior_body["record_type"] != "decision" or suspension is None:
                    errors.append(f"resolution_prior_not_suspended:{receipt_id}")
                else:
                    if suspension["suspension_id"] != resolution["suspension_id"]:
                        errors.append(f"resolution_suspension_mismatch:{receipt_id}")
                    if not _compatible_resolution(suspension, resolution["method"]):
                        errors.append(f"resolution_method_mismatch:{receipt_id}")
                    if resolution["resolved_at_ms"] < prior_body["timestamp_ms"]:
                        errors.append(f"resolution_time_invalid:{receipt_id}")
                    if (
                        resolution["method"]
                        in ("approval_granted", "context_supplied")
                        and resolution["resolved_at_ms"]
                        > suspension["expires_at_ms"]
                    ):
                        errors.append(f"resolution_after_expiry:{receipt_id}")
                if not _same_action(prior_body["action"], body["action"]):
                    errors.append(f"resolution_action_mismatch:{receipt_id}")
                if prior_body["initiator"] != body["initiator"]:
                    errors.append(f"resolution_initiator_mismatch:{receipt_id}")
        by_hash[current["receipt_hash"]] = current
        previous = current
    return {"valid": not errors, "errors": errors}

File: sdk-typescript/src/governance/strict-receipt-verify.ts
import { createHash, createPublicKey, verify as cryptoVerify } from 'node:crypto';
import { canonicalJsonForHash } from '../policy/tool-pinning.js';
import {
  STRICT_RECEIPT_BODY_DOMAIN,
  STRICT_RECEIPT_ENVELOPE_SCHEMA,
  buildStrictReceiptBody,
  strictReceiptKeyId,
  strictReceiptSignaturePreimage,
  type StrictReceiptBody,
  type StrictReceiptEnvelope,
} from './strict-receipt.js';

const HEX64 = /^[0-9a-f]{64}$/;
const HEX128 = /^[0-9a-f]{128}$/;
const KEY_ID = /^sha256:[0-9a-f]{64}$/;
const SPKI_PREFIX = Buffer.from('302a300506032b6570032100', 'hex');

export type StrictKeyTrust = 'pinned' | 'registered' | 'revoked' | 'self_asserted' | 'unknown';

export interface StrictReceiptVerification {
  schema_valid: boolean;
  hash_valid: boolean;
  signature_valid: boolean;
  semantic_valid: boolean;
  identity_binding_valid: boolean;
  key_trust: StrictKeyTrust;
}

function record(value: unknown): Record<string, unknown> | undefined {
  return value !== null && typeof value === 'object' && !Array.isArray(value)
    ? value as Record<string, unknown>
    : undefined;
}

function exact(value: Record<string, unknown>, allowed: readonly string[]): boolean {
  const keys = new Set(allowed);
  return Object.keys(value).every((key) => keys.has(key));
}

function u64(value: number): Buffer {
  const output = Buffer.alloc(8);
  output.writeBigUInt64BE(BigInt(value));
  return output;
}

function decodePublicKey(value: unknown): Buffer | undefined {
  if (typeof value !== 'string') return undefined;
  try {
    const raw = Buffer.from(value, 'base64');
    return raw.length === 32 && raw.toString('base64') === value ? raw : undefined;
  } catch { return undefined; }
}

export function verifyStrictReceipt(
  value: unknown,
  options: { pinned_public_key_b64?: string } = {},
): StrictReceiptVerification {
  const envelope = record(value);
  const body = record(envelope?.body);
  const signature = record(envelope?.signature);
  const schemaValid = envelope !== undefined
    && exact(envelope, ['schema', 'body', 'receipt_hash', 'signature', 'public_key_b64'])
    && envelope.schema === STRICT_RECEIPT_ENVELOPE_SCHEMA
    && body !== undefined
    && typeof envelope.receipt_hash === 'string' && HEX64.test(envelope.receipt_hash)
    && signature !== undefined
    && exact(signature, ['algorithm', 'key_id', 'value'])
    && signature.algorithm === 'Ed25519'
    && typeof signature.key_id === 'string' && KEY_ID.test(signature.key_id)
    && typeof signature.value === 'string' && HEX128.test(signature.value)
    && (!Object.prototype.hasOwnProperty.call(envelope, 'public_key_b64')
      || typeof envelope.public_key_b64 === 'string');

  let semanticValid = false;
  if (body) {
    try { buildStrictReceiptBody(body as unknown as StrictReceiptBody); semanticValid = true; } catch { /* false */ }
  }

  let hashValid = false;
  if (body && envelope && typeof envelope.receipt_hash === 'string') {
    try {
      const bytes = Buffer.from(canonicalJsonForHash(body), 'utf8');
      hashValid = createHash('sha256').update(Buffer.concat([
        Buffer.from(STRICT_RECEIPT_BODY_DOMAIN, 'utf8'), Buffer.from([0]),
        u64(bytes.length), bytes,
      ])).digest('hex') === envelope.receipt_hash;
    } catch { /* false */ }
  }

  const pinSupplied = Object.prototype.hasOwnProperty.call(options, 'pinned_public_key_b64');
  const pinned = decodePublicKey(options.pinned_public_key_b64);
  const embedded = decodePublicKey(envelope?.public_key_b64);
  const rawKey = pinSupplied ? pinned : embedded;
  const keyTrust: StrictKeyTrust = pinSupplied ? 'pinned' : embedded ? 'self_asserted' : 'unknown';
  const initiator = record(body?.initiator);
  const bodyKeyId = initiator?.key_id;
  const signatureKeyId = signature?.key_id;
  const fingerprint = rawKey ? strictReceiptKeyId(rawKey) : undefined;
  const identityValid = typeof bodyKeyId === 'string'
    && typeof signatureKeyId === 'string'
    && fingerprint !== undefined
    && bodyKeyId === signatureKeyId
    && signatureKeyId === fingerprint;

  let signatureValid = false;
  if (rawKey && typeof signatureKeyId === 'string' && envelope
    && typeof envelope.receipt_hash === 'string' && typeof signature?.value === 'string') {
    try {
      const key = createPublicKey({
        key: Buffer.concat([SPKI_PREFIX, rawKey]), format: 'der', type: 'spki',
      });
      signatureValid = cryptoVerify(
        null,
        strictReceiptSignaturePreimage(signatureKeyId, envelope.receipt_hash),
        key,
        Buffer.from(signature.value, 'hex'),
      );
    } catch { /* false */ }
  }

  return {
    schema_valid: schemaValid,
    hash_valid: hashValid,
    signature_valid: signatureValid,
    semantic_valid: semanticValid,
    identity_binding_valid: identityValid,
    key_trust: keyTrust,
  };
}

function sameAction(left: StrictReceiptBody['action'], right: StrictReceiptBody['action']): boolean {
  return left.action_id === right.action_id
    && left.kind === right.kind
    && left.name === right.name
    && left.arguments_hash === right.arguments_hash
    && left.target === right.target;
}

function compatibleResolution(
  suspension: NonNullable<StrictReceiptBody['suspension']>,
  method: NonNullable<StrictReceiptBody['resolution']>['method'],
): boolean {
  if (suspension.type === 'approval') {
    return ['approval_granted', 'approval_denied', 'expired', 'cancelled'].includes(method);
  }
  return ['context_supplied', 'expired', 'cancelled'].includes(method);
}

export function verifyStrictReceiptChain(
  envelopes: readonly unknown[],
  options: { pinned_public_key_b64?: string } = {},
): { valid: boolean; errors: string[] } {
  if (envelopes.length === 0) return { valid: false, errors: ['empty_chain'] };
  const errors: string[] = [];
  const byHash = new Map<string, StrictReceiptEnvelope>();
  const resolved = new Set<string>();
  let session: string | undefined;
  let previous: StrictReceiptEnvelope | undefined;
  for (let index = 0; index < envelopes.length; index++) {
    const candidate = envelopes[index];
    const rawEnvelope = record(candidate);
    const rawBody = record(rawEnvelope?.body);
    const rawId = rawBody?.receipt_id;
    const id = typeof rawId === 'string' ? rawId : `index-${index}`;
    const axes = verifyStrictReceipt(candidate, options);
    if (!axes.schema_valid) errors.push(`receipt_schema_invalid:${id}`);
    if (!axes.semantic_valid) errors.push(`receipt_semantic_invalid:${id}`);
    if (!axes.hash_valid) errors.push(`receipt_hash_invalid:${id}`);
    if (!axes.signature_valid) errors.push(`receipt_signature_invalid:${id}`);
    if (!axes.identity_binding_valid) errors.push(`receipt_identity_invalid:${id}`);
    if (!axes.schema_valid || !axes.semantic_valid || !axes.hash_valid
      || !axes.signature_valid || !axes.identity_binding_valid || !rawEnvelope || !rawBody) {
      continue;
    }
    const current = {
      ...rawEnvelope,
      body: buildStrictReceiptBody(rawBody as unknown as StrictReceiptBody),
    } as unknown as StrictReceiptEnvelope;
    if (session === undefined) session = current.body.session_id;

    if (current.body.sequence !== index + 1) errors.push(`sequence_order_invalid:${id}`);
    if (current.body.session_id !== session) errors.push(`session_mismatch:${id}`);
    if (previous) {
      if (current.body.previous_receipt_hash !== previous.receipt_hash) errors.push(`previous_hash_mismatch:${id}`);
      if (current.body.timestamp_ms < previous.body.timestamp_ms) errors.push(`timestamp_regression:${id}`);
    }

    const resolution = current.body.resolution;
    if (current.body.record_type === 'resolution' && resolution) {
      if (resolved.has(resolution.resolves_receipt_hash)) {
        errors.push(`duplicate_resolution:${id}`);
      } else {
        resolved.add(resolution.resolves_receipt_hash);
      }
      const prior = byHash.get(resolution.resolves_receipt_hash);
      if (!prior) {
        errors.push(`resolution_reference_invalid:${id}`);
      } else {
        const suspension = prior.body.suspension;
        if (prior.body.record_type !== 'decision' || !suspension) {
          errors.push(`resolution_prior_not_suspended:${id}`);
        } else {
          if (suspension.suspension_id !== resolution.suspension_id) errors.push(`resolution_suspension_mismatch:${id}`);
          if (!compatibleResolution(suspension, resolution.method)) errors.push(`resolution_method_mismatch:${id}`);
          if (resolution.resolved_at_ms < prior.body.timestamp_ms) errors.push(`resolution_time_invalid:${id}`);
          if ((resolution.method === 'approval_granted' || resolution.method === 'context_supplied')
            && resolution.resolved_at_ms > suspension.expires_at_ms) {
            errors.push(`resolution_after_expiry:${id}`);
          }
        }
        if (!sameAction(prior.body.action, current.body.action)) errors.push(`resolution_action_mismatch:${id}`);
        if (prior.body.initiator.agent_id !== current.body.initiator.agent_id
          || prior.body.initiator.key_id !== current.body.initiator.key_id) {
          errors.push(`resolution_initiator_mismatch:${id}`);
        }
      }
    }
    byHash.set(current.receipt_hash, current);
    previous = current;
  }
  return { valid: errors.length === 0, errors };
}
